src/modules/0.0.11/aws_ebs/entity/general_purpose_volume.ts

```diff
@@ -0,0 +1,114 @@
+import {
+  Entity,
+  Column,
+  PrimaryGeneratedColumn,
+  ManyToOne,
+  JoinColumn,
+  AfterLoad,
+  AfterInsert,
+  AfterUpdate,
+  Check,
+  Unique,
+} from 'typeorm'
+
+import { cloudId, } from '../../../../services/cloud-id'
+import { Instance } from '../../aws_ec2/entity';
+import { AvailabilityZone } from '../../aws_vpc/entity';
+
+export enum GeneralPurposeVolumeType {
+  GP2 = "gp2",
+  GP3 = "gp3",
+}
+
+export enum VolumeState {
+  AVAILABLE = "available",
+  CREATING = "creating",
+  DELETED = "deleted",
+  DELETING = "deleting",
+  ERROR = "error",
+  IN_USE = "in-use",
+}
+
+@Unique('Unique_gp_instance_device_name', ['instanceDeviceName', 'attachedInstance'])
+@Entity()
+export class GeneralPurposeVolume {
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @Column({ nullable: true, })
+  @cloudId
+  volumeId?: string;
+
+  @Column({
+    nullable: false,
+    type: 'enum',
+    enum: GeneralPurposeVolumeType,
+  })
+  volumeType: GeneralPurposeVolumeType;
+
+  @Column({
+    type: 'enum',
+    enum: AvailabilityZone,
+  })
+  availabilityZone: AvailabilityZone;
+
+  @Check('Check_gp_volume_size_min_max', `"size" > 0 AND "size" < 16385`)
+  @Column({
+    type: 'int',
+    default: 8,  // AWS default when creating an instance
+  })
+  size: number;
+
+  @Column({
+    nullable: true,
+    type: 'enum',
+    enum: VolumeState,
+  })
+  state?: VolumeState;
+
+  @ManyToOne(() => Instance, instance => instance.id, {
+    eager: true,
+    nullable: true,
+  })
+  @JoinColumn({ name: 'attached_instance_id', })
+  attachedInstance?: Instance;
```

Comment:
Interesting, I assumed the relationship here would be inverted and ec2 would own the fk relationship. This looks good too. I didn’t realize it was possible to attach multiple volumes / hard drives to a single instance, but I suppose that makes sense. Is it possible to attach a single volume to multiple instances? Not that we should support it on the first pass, mostly curious.

Reply:
No, you definitely can't attach a single volume to multiple instances. You can detach a volume from one and re-attach to another, but AWS won't auto-NFS these volumes or anything like that (since that's a completely different layer of the stack in most OSes and more importantly varies between OSes; they can't just emulate a hard drive on a PCI bus and be done with it).

Reply:
Actually, one specific type of volume (not the general purpose I've mapped here) can be attached to multiple instances, but I think is to a specific instance type (nitro instances or something like that).

That was one of the reason I've decided to split the volume types and here just mapped the general purpose ones.